### DISCUSSION TITLE
FastAPI dependencies. Parameters hell.

### DISCUSSION THREAD

**Discussion by ph55** [fastapi/fastapi#5576] [Questions] [answered]
I'm trying to understand the use-case and benefits of FastAPI dependencies.

If I can use them only in "route functions", 
does it mean that I need to pass dependency as the parameter to every nested function.

Examples:

1. Application settings. 
It is recommended to use app settings/config as dependency:
https://fastapi.tiangolo.com/advanced/settings/#the-config-file
![image](https://user-images.githubusercontent.com/1386622/199472590-29e01ae5-f909-4e88-9412-0dba80282def.png)

Now, every library that should access the setting should accept it as the argument ?

2.  Application logging.
I have `AppLogger` dependency which relies on `request_id`. Each log message have `request_id` prefixed to it.
Request id generated per request by "RequestIdMiddleware".

Now again, every library should receive the logger as the argument ?

**Comment by tiangolo** [comment, marked as answer]
It's probably better if you ask in an issue, that will help guide you to structure the question in a way that is easier to understand what you need and help you, also sharing a minimal code example.

**Comment by shreve** [comment]
Class-based views are really helpful for this. I'd like to see them upstream in FastAPI, but for now, there's a 3rd party library, [fastapi-restful](https://fastapi-restful.netlify.app/user-guide/class-based-views/) which implements them. 

```python
from fastapi import APIRouter, Depends
from fastapi_restful.cbv import cbv

router = APIRouter()

class AppView:
    log: AppLogger = Depends(get_app_logger)

@cbv(router)
class InfoView(AppView):
    @router.get("/info")
    async def info(self):
        self.log.info("This dependency is injected for you without explicit definition in the function sig")
```
